source/vsignal/slot.d

```diff
@@ -0,0 +1,163 @@
+module vsignal.slot;
+
+import std.traits : ReturnType, Parameters;
+
+void connect(alias pred, F)(auto ref Slot!F slot)
+{
+	import vsignal.utils : tryForward;
+
+	with(slot)
+	{
+		payload = null;
+		fn = (const scope void*, Parameters args) {
+			static if (is(ReturnType == void))
+				pred(tryForward!(pred, args));
+			else
+				return ReturnType(pred(tryForward!(pred, args)));
+		};
+	}
+}
+
+@safe pure nothrow @nogc unittest
+{
+	Slot!(void delegate(int)) slot;
+
+	class C
+	{
+		static void opCall(int) {}
+		static void fooA(const int) {}
+		static void fooB(ref int) {}
+		static void fooC(const ref int) {}
+	}
+
+	// all lambda listeners can omit the type
+	slot.connect!((ref _)       {});
+	slot.connect!((const ref _) {});
+	slot.connect!((_)           {});
+	slot.connect!((const _)     {});
+
+	slot.connect!(C.opCall);
+	slot.connect!(C.fooA);
+	slot.connect!(C.fooB);
+	slot.connect!(C.fooC);
+}
+
+void connect(alias pred, T, F)(auto ref Slot!F slot, ref T instance)
+{
+	import vsignal.utils : tryForward;
+	import std.format : format;
+	import std.traits : isCallable;
+
+	static immutable id = __traits(identifier, pred);
+	static immutable Tinit = T.init;
+
+	// check is pred is (ref T, ...)
+	static if (isCallable!pred && is(Parameters!pred[0] == T))
+		enum str = q{ pred(*() @trusted { return cast(T*) payload; } (), tryForward!(pred, Tinit, args)[1 .. $]) };
+
+	// check is pred is a method of intance
+	else static if (__traits(hasMember, instance, id))
+		enum str = q{ () @trusted { return cast(T*) payload; } ().%s(tryForward!(pred, args)) }.format(id);
+
+	// fallback to any possible function
+	else
+		enum str = q{ pred(*() @trusted { return cast(T*) payload; } (), tryForward!(pred, Tinit, args)[1 .. $]) };
+
+
+	with(slot)
+	{
+		() @trusted { payload = cast(void*) &instance; } ();
+		fn = (const scope void* payload, Parameters args) {
+			static if (is(ReturnType == void))
+				() @trusted { cast(void) mixin(str); } ();
```

Comment:
That actually needs to be reworked. The first @trusted is to work when converting for example `immutable(int*)` to `void*`. The instance cannot be implicitly converted in those cases. The problem is that it breaks the whole `cannot pass the reference of a locally defined type in @safe code` error.

The second @trusted is just to imply the data is discarded, which in that case I already guarantee it is, so kind of unnecessary.

Reply:
This will probably solve the first issue:

```suggestion
- () @trusted { payload = cast(void*) &instance; } ();
+ payload = () @trusted { cast(Unqual!T*) &instance; } ();
```

Reply:
done